pkg/kubelet/images/pullmanager/image_pull_policies.go

```diff
@@ -0,0 +1,151 @@
+// for the `image` exist on the node, meaning it was pulled by the kubelet somewhere
+// in the past.
+type ImagePullPolicyEnforcer func(image string, imagePresent, imagePulledByKubelet bool) bool
+
+func NewImagePullCredentialVerificationPolicy(policy kubeletconfiginternal.ImagePullCredentialsVerificationPolicy, imageAllowList []string) (ImagePullPolicyEnforcer, error) {
+	switch policy {
+	case kubeletconfiginternal.NeverVerify:
+		return ImagePullPolicyEnforcer(NeverVerifyImagePullPolicy), nil
+	case "", kubeletconfiginternal.NeverVerifyPreloadedImages:
+		return ImagePullPolicyEnforcer(NeverVerifyPreloadedPullPolicy), nil
+	case kubeletconfiginternal.NeverVerifyAllowlistedImages:
+		return NewNeverVerifyAllowListedPullPolicy(imageAllowList)
+	case kubeletconfiginternal.AlwaysVerify:
+		return ImagePullPolicyEnforcer(AlwaysVerifyImagePullPolicy), nil
+	default:
+		return nil, fmt.Errorf("unknown image pull credential verification policy: %v", policy)
+	}
+}
+
+func NeverVerifyImagePullPolicy(image string, imagePresent, imagePulledByKubelet bool) bool {
+	return false
+}
+
+func NeverVerifyPreloadedPullPolicy(image string, imagePresent, imagePulledByKubelet bool) bool {
+	if imagePresent && !imagePulledByKubelet {
+		return false
+	}
+	return true
+}
+
+func AlwaysVerifyImagePullPolicy(image string, imagePresent, imagePulledByKubelet bool) bool {
+	return true
+}
+
+type NeverVerifyAllowlistedImages struct {
+	absoluteURLs sets.Set[string]
+	prefixes     []string
+}
+
+func (p *NeverVerifyAllowlistedImages) RequiresVerification(image string, imagePresent, imagePulledByKubelet bool) bool {
+	if imagePresent && p.imageMatches(image) {
+		return false
+	}
+	return true
+}
+
+func (p *NeverVerifyAllowlistedImages) imageMatches(image string) bool {
+	if p.absoluteURLs.Has(image) {
+		return true
+	}
+	for _, prefix := range p.prefixes {
+		if strings.HasPrefix(image, prefix) {
+			return true
+		}
+	}
+	return false
+}
+
+func (p *NeverVerifyAllowlistedImages) AllowPattern(pattern string) error {
+	pattern = strings.TrimSpace(pattern)
+
+	if err := validateAllowlistImagePattern(pattern); err != nil {
+		return err
+	}
+
+	if strings.HasSuffix(pattern, "/*") {
+		p.prefixes = append(p.prefixes, pattern[:len(pattern)-2])
+	} else {
+		p.absoluteURLs.Insert(pattern)
+	}
+	return nil
+}
+
+func NewNeverVerifyAllowListedPullPolicy(allowList []string) (ImagePullPolicyEnforcer, error) {
+	policy := &NeverVerifyAllowlistedImages{
+		absoluteURLs: sets.New[string](),
+	}
+	for _, pattern := range allowList {
+		if err := policy.AllowPattern(pattern); err != nil {
+			return nil, err
+		}
+	}
+
+	return policy.RequiresVerification, nil
+}
+
+func ValidateAllowlistImagesPatterns(patterns []string) error {
+	for _, p := range patterns {
+		if err := validateAllowlistImagePattern(p); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func validateAllowlistImagePattern(pattern string) error {
+	if strings.ContainsRune(pattern[:len(pattern)-1], '*') {
```

Comment:
does this panic if the pattern is ""?

Reply:
simplify to avoid possible index panics, etc

```go
func getTrimmedPattern(pattern string) (string, bool, error) {
  trimmedPattern := ""
  isPrefixPattern = false
  if strings.HasSuffix(pattern, "/*") {
    trimmed = strings.TrimSuffix(pattern, "*")
    isPrefixPattern = true
  } else {
    trimmedPattern = pattern
  }

  if len(trimmedPattern) == 0 {
    return error
  }
  if strings.ContainsRune(trimmedPattern, '*') {
    return error
  }
  if !isPrefixPattern && strings.HasSuffix(trimmedPattern, "/") {
    return error
  }
  return nil
}
```